File: app/models/referral.py
"""Retained historical CRM referral mirror rows (RFC #73).

Native referral tables are authoritative. These legacy rows are read-only
migration/retention evidence: no customer surface, webhook, task, or service
hydrates them or derives a native referral decision from them.
"""

import uuid
from datetime import UTC, datetime
from decimal import Decimal

from sqlalchemy import ForeignKey, Numeric, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.types import DateTime

from app.db import Base


class ReferralMirror(Base):
    """One retained historical CRM referral snapshot."""

    __tablename__ = "referral_mirror"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    crm_referral_id: Mapped[str] = mapped_column(
        String(64), nullable=False, unique=True, index=True
    )
    subscriber_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("subscribers.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    referred_name: Mapped[str | None] = mapped_column(String(160))
    # pending | qualified | rewarded | rejected (CRM ReferralStatus values).
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="pending")
    reward_amount: Mapped[Decimal | None] = mapped_column(Numeric(12, 2))
    reward_currency: Mapped[str] = mapped_column(
        String(3), nullable=False, default="NGN"
    )
    # none | pending | approved | paid (CRM ReferralRewardStatus values).
    reward_status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="none"
    )
    referral_created_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True)
    )
    qualified_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    rewarded_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(UTC)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(UTC),
        onupdate=lambda: datetime.now(UTC),
    )


class ReferralProgramCache(Base):
    """Retained historical per-subscriber referral program snapshot."""

    __tablename__ = "referral_program_cache"

    subscriber_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("subscribers.id", ondelete="CASCADE"),
        primary_key=True,
    )
    code: Mapped[str] = mapped_column(String(32), nullable=False)
    share_url: Mapped[str] = mapped_column(String(255), nullable=False)
    program_enabled: Mapped[bool] = mapped_column(nullable=False, default=True)
    reward_amount: Mapped[Decimal | None] = mapped_column(Numeric(12, 2))
    reward_currency: Mapped[str] = mapped_column(
        String(3), nullable=False, default="NGN"
    )
    synced_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(UTC),
        onupdate=lambda: datetime.now(UTC),
    )

File: mobile/lib/src/models/referral.dart
// Refer & Earn models — mirror of the sub `/me/referrals` payload (RFC #73).

class ReferralProgram {
  ReferralProgram({
    required this.enabled,
    required this.rewardAmount,
    required this.rewardCurrency,
  });

  final bool enabled;
  final String rewardAmount; // decimal string, e.g. "5000.00"
  final String rewardCurrency;

  factory ReferralProgram.fromJson(Map<String, dynamic> json) =>
      ReferralProgram(
        enabled: json['enabled'] == true,
        rewardAmount: (json['reward_amount'] ?? '0').toString(),
        rewardCurrency: (json['reward_currency'] ?? 'NGN').toString(),
      );
}

class ReferralTotals {
  ReferralTotals({
    this.total = 0,
    this.pending = 0,
    this.qualified = 0,
    this.rewarded = 0,
    this.totalEarned = '0',
  });

  final int total;
  final int pending;
  final int qualified;
  final int rewarded;
  final String totalEarned;

  factory ReferralTotals.fromJson(Map<String, dynamic> json) => ReferralTotals(
    total: _asInt(json['total']),
    pending: _asInt(json['pending']),
    qualified: _asInt(json['qualified']),
    rewarded: _asInt(json['rewarded']),
    totalEarned: (json['total_earned'] ?? '0').toString(),
  );
}

class ReferralItem {
  ReferralItem({
    required this.id,
    required this.status,
    required this.rewardCurrency,
    required this.rewardStatus,
    this.referredName,
    this.rewardAmount,
    this.createdAt,
  });

  final String id;
  final String status; // pending | qualified | rewarded | rejected
  final String? referredName;
  final String? rewardAmount;
  final String rewardCurrency;
  final String rewardStatus;
  final DateTime? createdAt;

  factory ReferralItem.fromJson(Map<String, dynamic> json) => ReferralItem(
    id: json['id'].toString(),
    status: (json['status'] ?? 'pending').toString(),
    referredName: json['referred_name'] as String?,
    rewardAmount: json['reward_amount']?.toString(),
    rewardCurrency: (json['reward_currency'] ?? 'NGN').toString(),
    rewardStatus: (json['reward_status'] ?? 'none').toString(),
    createdAt: _asDate(json['created_at']),
  );
}

class ReferralSummary {
  ReferralSummary({
    required this.code,
    required this.shareUrl,
    required this.program,
    required this.totals,
    required this.referrals,
  });

  final String code;
  final String shareUrl;
  final ReferralProgram program;
  final ReferralTotals totals;
  final List<ReferralItem> referrals;

  factory ReferralSummary.fromJson(Map<String, dynamic> json) =>
      ReferralSummary(
        code: (json['code'] ?? '').toString(),
        shareUrl: (json['share_url'] ?? '').toString(),
        program: ReferralProgram.fromJson(
          (json['program'] as Map<String, dynamic>?) ?? const {},
        ),
        totals: ReferralTotals.fromJson(
          (json['totals'] as Map<String, dynamic>?) ?? const {},
        ),
        referrals: ((json['referrals'] as List?) ?? const [])
            .whereType<Map<String, dynamic>>()
            .map(ReferralItem.fromJson)
            .toList(),
      );
}

int _asInt(dynamic v) {
  if (v is int) return v;
  return int.tryParse(v?.toString() ?? '') ?? 0;
}

DateTime? _asDate(dynamic v) {
  if (v == null) return null;
  return DateTime.tryParse(v.toString())?.toLocal();
}
